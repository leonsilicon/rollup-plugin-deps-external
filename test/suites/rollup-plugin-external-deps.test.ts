import commonjs from '@rollup/plugin-commonjs';
import nodeResolve from '@rollup/plugin-node-resolve';
import { join } from 'desm';
import * as path from 'node:path';
import type { RollupOptions } from 'rollup';
import { rollup } from 'rollup';
import { describe, expect, it } from 'vitest';

import externalDeps from '~/index.js';
import type { ExternalDepsPluginOptions } from '~/types.js';

async function bundle(
	{ input, external }: { input: string; external?: RollupOptions['external'] },
	options?: ExternalDepsPluginOptions
) {
	return rollup({
		external,
		input,
		plugins: [externalDeps(options), nodeResolve(), commonjs()],
	})
		.then(async (bundle) => bundle.generate({ format: 'esm' }))
		.then((result) => result.output[0].code)
		.then((code) => {
			expect(code).toMatchSnapshot();
		});
}

describe('autoExternal(options)', () => {
	it('should have a name', () => {
		expect(externalDeps().name).toEqual('external-deps');
	});

	it('should add dependencies by default', async () =>
		bundle({
			input: join(import.meta.url, '../fixtures/deps/index.js'),
		}));

	it('should handle disabling dependencies', async () =>
		bundle(
			{ input: join(import.meta.url, '../fixtures/deps/index.js') },
			{ dependencies: false }
		));

	it('should add peerDependencies by default', async () =>
		bundle({
			input: join(import.meta.url, '../fixtures/peer-deps/index.js'),
		}));

	it('should handle disabling peerDependencies', async () =>
		bundle(
			{ input: join(import.meta.url, '../fixtures/peer-deps/index.js') },
			{ peerDependencies: false }
		));

	it('should add builtins by default', async () =>
		bundle({
			input: join(import.meta.url, '../fixtures/builtins/index.js'),
		}));

	it('should handle disabling builtins', async () =>
		bundle(
			{ input: join(import.meta.url, '../fixtures/builtins/index.js') },
			{ builtins: false }
		));

	it('should handle adding builtins for a specific Node.js version', async () =>
		bundle({
			input: join(import.meta.url, '../fixtures/builtins-6.0.0/index.js'),
		}));

	it('should handle extending external array', async () =>
		bundle({
			input: join(import.meta.url, '../fixtures/deps/index.js'),
			external: ['baz'],
		}));

	it('should handle extending external function', async () =>
		bundle({
			input: join(import.meta.url, '../fixtures/deps/index.js'),
			external: (id) => id.includes('baz'),
		}));

	it('should handle a custom package file path', async () =>
		bundle(
			{ input: join(import.meta.url, '../fixtures/deps/index.js') },
			{
				packagePath: join(
					import.meta.url,
					'../fixtures/package-path/package.json'
				),
			}
		));

	it('should handle a directory path as the packagePath', async () =>
		bundle(
			{ input: join(import.meta.url, '../fixtures/deps/index.js') },
			{ packagePath: join(import.meta.url, '../fixtures/package-path') }
		));

	it('should handle cherry picked modules', async () =>
		bundle({
			input: join(import.meta.url, '../fixtures/cherry-picked/index.js'),
		}));

	it('should handle scoped modules', async () =>
		bundle({
			input: join(import.meta.url, '../fixtures/scoped/index.js'),
		}));
});
