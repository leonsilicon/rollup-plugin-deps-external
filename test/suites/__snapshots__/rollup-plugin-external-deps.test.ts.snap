// Vitest Snapshot v1

exports[`autoExternal(options) > should add builtins by default 1`] = `
"import require$$0 from 'path';

const path = require$$0;

var builtins = path;

export { builtins as default };
"
`;

exports[`autoExternal(options) > should add dependencies by default 1`] = `
"var foo = 'foo';

var bar$1 = 'bar';

const bar = bar$1;

var bar_1 = bar;

var baz = 'baz';

var index = \`\${foo}\${bar_1}\${baz}\`;

export { index as default };
"
`;

exports[`autoExternal(options) > should add peerDependencies by default 1`] = `
"var foo = 'foo';

var bar$1 = 'bar';

const bar = bar$1;

var bar_1 = bar;

var baz = 'baz';

var index = \`\${foo}\${bar_1}\${baz}\`;

export { index as default };
"
`;

exports[`autoExternal(options) > should handle a custom package file path 1`] = `
"var foo = 'foo';

var bar$1 = 'bar';

const bar = bar$1;

var bar_1 = bar;

var baz = 'baz';

var index = \`\${foo}\${bar_1}\${baz}\`;

export { index as default };
"
`;

exports[`autoExternal(options) > should handle a directory path as the packagePath 1`] = `
"var foo = 'foo';

var bar$1 = 'bar';

const bar = bar$1;

var bar_1 = bar;

var baz = 'baz';

var index = \`\${foo}\${bar_1}\${baz}\`;

export { index as default };
"
`;

exports[`autoExternal(options) > should handle adding builtins for a specific Node.js version 1`] = `
"import require$$0 from 'os';

const os = require$$0;

var builtins6_0_0 = os;

export { builtins6_0_0 as default };
"
`;

exports[`autoExternal(options) > should handle cherry picked modules 1`] = `
"var foo = 'foo';

var bar = 'bar';

var baz = 'baz';

var index = \`\${foo}\${bar}\${baz}\`;

export { index as default };
"
`;

exports[`autoExternal(options) > should handle disabling builtins 1`] = `
"import require$$0 from 'path';

const path = require$$0;

var builtins = path;

export { builtins as default };
"
`;

exports[`autoExternal(options) > should handle disabling dependencies 1`] = `
"var foo = 'foo';

var bar$1 = 'bar';

const bar = bar$1;

var bar_1 = bar;

var baz = 'baz';

var index = \`\${foo}\${bar_1}\${baz}\`;

export { index as default };
"
`;

exports[`autoExternal(options) > should handle disabling peerDependencies 1`] = `
"var foo = 'foo';

var bar$1 = 'bar';

const bar = bar$1;

var bar_1 = bar;

var baz = 'baz';

var index = \`\${foo}\${bar_1}\${baz}\`;

export { index as default };
"
`;

exports[`autoExternal(options) > should handle extending external array 1`] = `
"import baz from 'baz';

var foo = 'foo';

var bar$1 = 'bar';

const bar = bar$1;

var bar_1 = bar;

var index = \`\${foo}\${bar_1}\${baz}\`;

export { index as default };
"
`;

exports[`autoExternal(options) > should handle extending external function 1`] = `
"import baz from 'baz';

var foo = 'foo';

var bar$1 = 'bar';

const bar = bar$1;

var bar_1 = bar;

var index = \`\${foo}\${bar_1}\${baz}\`;

export { index as default };
"
`;

exports[`autoExternal(options) > should handle scoped modules 1`] = `
"var foo = '@scope/foo';

var bar$1 = '@scope/bar';

const bar = bar$1;

var bar_1 = bar;

var baz = '@scope/baz';

var index = \`\${foo}\${bar_1}\${baz}\`;

export { index as default };
"
`;
